README.md
# ShortIo

[![Gem Version](https://badge.fury.io/rb/short_io.svg)](https://badge.fury.io/rb/short_io) | [![Build Status](https://app.travis-ci.com/yosefbennywidyo/short_io.svg?branch=main)](https://app.travis-ci.com/yosefbennywidyo/short_io) | [![codecov](https://codecov.io/gh/yosefbennywidyo/short_io/branch/main/graph/badge.svg?token=MBMxFB57mZ)](https://codecov.io/gh/yosefbennywidyo/short_io)

A Ruby gem to use with [short.io](https://short.io)

[ShortIo gem documentation](https://rubydoc.info/gems/short_io)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'short_io'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install short_io

## Usage

```ruby
require 'short_io'

ShortIo::ShortUrl.new('kreasi-guna-semesta.co.id', 'API_KEY')
```

### Add Domain

```ruby
ShortIo::ShortUrl.new('kreasi-guna-semesta.co.id', 'API_KEY').add_domain
```

### List Domain

```ruby
ShortIo::ShortUrl.new('kreasi-guna-semesta.co.id', 'API_KEY').domain_list
```

### Update Domain

```ruby
ShortIo::ShortUrl.new('example.com', 'YOUR_API_KEY', {domain_id: YOUR_DOMAIN_ID, root_redirect_url: 'YOUR_LINK'}).update_domain
```

### Delete Domain

```ruby
ShortIo::ShortUrl.new('example.com', 'YOUR_API_KEY', {domain_id: YOUR_DOMAIN_ID}).delete_domain
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/yosefbennywidyo/short_io. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/yosefbennywidyo/short_io/blob/main/CODE_OF_CONDUCT.md).


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the ShortIoShortBrandedUrl project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/yosefbennywidyo/short_io/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
0.1.10 (2022-02-10)
------------------

* add delete and update method
* initialize using four (4) parameters
* setup method using two (2) default parameters
* add documentation for domain_list, update, and delete
* update version to [0.1.10](https://rubygems.org/gems/short_io/versions/0.1.10)

0.1.9 (2022-02-03)
------------------

* add documentation
* update version to [0.1.9](https://rubygems.org/gems/short_io/versions/0.1.9)

0.1.8 (2022-02-03)
------------------

* add webmock gem
* add test for add_domain and domain list methods
* add `:host_name`, `:api_key`, and `:options` attr_reader
* update version to [0.1.8](https://rubygems.org/gems/short_io/versions/0.1.8)

0.1.7 (2022-01-25)
------------------

* add test
* add memoization
* add check_variables method to raise error when `nil` or `empty` on host name or api_key
* update version to [0.1.7](https://rubygems.org/gems/short_io/versions/0.1.7)

0.1.6 (2022-01-25)
------------------

* fix version on CHANGELOG.md
* update version to [0.1.6](https://rubygems.org/gems/short_io/versions/0.1.6)

0.1.5 (2022-01-25)
------------------

* fix CHANGELOG.md path
* update version to [0.1.5](https://rubygems.org/gems/short_io/versions/0.1.5)

0.1.4 (2022-01-25)
------------------

* add CHANGELOG.md
* update version to [0.1.4](https://rubygems.org/gems/short_io/versions/0.1.4)

0.1.3 (2022-01-25)
------------------

* add Gemfile.lock
* update version to [0.1.3](https://rubygems.org/gems/short_io/versions/0.1.3)

0.1.2 (2022-01-25)
------------------

* fix sample usage on README.md
* update version to [0.1.2](https://rubygems.org/gems/short_io/versions/0.1.2)

0.1.1 (2022-01-25)
------------------

* update version to [0.1.1](https://rubygems.org/gems/short_io/versions/0.1.1) (no changes)

0.1.0 (2022-01-25)
------------------

* [YANKED](https://rubygems.org/gems/short_io/versions/0.1.0) Start create `short_io` gem

test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require 'simplecov'
SimpleCov.start

require 'codecov'
SimpleCov.formatter = SimpleCov::Formatter::Codecov

require "short_io"

require "minitest/autorun"
require "webmock/minitest"

test/short_io_test.rb
require "test_helper"

class ShortIoTest < Minitest::Test
  SHORT_IO_BASE_URL       = 'https://api.short.io/domains/'
  DEFAULT_HOST_NAME       = 'example.com'
  DEFAULT_API_KEY         = 'API_KEY'
  SECOND_EXAMPLE_DOMAIN   = 'example-2.com'
  SECOND_EXAMPLE_API_KEY  = 'API_KEY_2'
  SUCCESS_DELETE_OR_UPDATE= "{
    success: true
  }"

  SUCCESS_ADD_DOMAIN      = "{
    linkType: 'random',
    state: 'configured',
    cloaking: false,
    setupType: 'dns',
    httpsLinks: false,
    id: 91576,
    hostname: 'yourdomain.com',
    UserId: 9346,
    updatedAt: '2020-04-23T10:22:47.010Z',
    createdAt: '2020-04-23T10:22:46.649Z',
    provider: null,
    unicodeHostname: 'urdomain.com',
    clientStorage: null
  }"

  SUCCESS_DOMAIN_LIST      = "{
    id: 7252,
    hostname: 'yrbrand.co',
    title: null,
    segmentKey: null,
    linkType: 'increment',
    state: 'not_configured',
    provider: 'cloudflare',
    redirect404: 'https://short.cm',
    hideReferer: 1,
    caseSensitive: true,
    exportEnabled: true,
    cloaking: false,
    jsRedir: true,
    incrementCounter: 'A',
    setupType: 'js',
    autodeletePeriod: 1,
    httpsLinks: true,
    clientStorage: '{'configurationHidden':false}',
    integrationGA: null,
    integrationFB: null,
    integrationAdroll: null,
    integrationGTM: null,
    createdAt: '2017-12-07T08:24:41.000Z',
    updatedAt: '2019-12-24T13:08:30.000Z',
    TeamId: 1381,
    unicodeHostname: 'yrbrand.co'
  }"

  def test_that_it_has_latest_version_number
    assert_match ::ShortIo::VERSION, '0.1.10'
  end

  def test_it_initialize_short_url_correctly
    default_initialize
  end

  def test_it_initialize_with_new_host_name
    default_initialize(SECOND_EXAMPLE_DOMAIN)
  end

  def test_it_initialize_with_new_api_key
    default_initialize(DEFAULT_HOST_NAME, SECOND_EXAMPLE_API_KEY)
  end

  def test_it_raise_error_when_nil_host_name
    assert_raises(ShortIo::HostNameError) { create_new_short_url(nil, SECOND_EXAMPLE_API_KEY) }
  end

  def test_it_raise_error_when_empty_host_name
    assert_raises(ShortIo::HostNameError) { create_new_short_url('', SECOND_EXAMPLE_API_KEY) }
  end

  def test_it_raise_error_when_nil_api_key
    assert_raises(ShortIo::ApiKeyError) { create_new_short_url(SECOND_EXAMPLE_DOMAIN, nil) }
  end

  def test_it_raise_error_when_empty_api_key
    assert_raises(ShortIo::ApiKeyError) { create_new_short_url(SECOND_EXAMPLE_DOMAIN, '') }
  end

  def test_it_initialize_with_new_options
    default_initialize(DEFAULT_HOST_NAME, DEFAULT_API_KEY, {hide_referer: true, https_link: true, link_type: 'long'})
  end

  def test_it_setup
    create_new_short_url(DEFAULT_HOST_NAME, DEFAULT_API_KEY)
    @short_url.setup
    @short_url.yield_self.inspect.include? '@url'
    @short_url.yield_self.inspect.include? SHORT_IO_BASE_URL
    @short_url.yield_self.inspect.include? '@http'
    @short_url.yield_self.inspect.include? 'api.short.io:443'
  end

  def test_it_add_domain
    success_add_domain
    create_new_short_url(DEFAULT_HOST_NAME, DEFAULT_API_KEY)
    @short_url.add_domain
    assert_equal @short_url.add_domain, SUCCESS_ADD_DOMAIN
  end

  def success_add_domain
    stub_request(:post, "https://api.short.io/domains/").
      with(
        body: "{\"hideReferer\":\"false\",\"httpsLinks\":\"false\",\"hostname\":\"example.com\",\"linkType\":\"random\"}",
        headers: {
              'Accept'=>'application/json',
              'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
              'Authorization'=>'API_KEY',
              'Content-Type'=>'application/json',
              'Host'=>'api.short.io',
              'User-Agent'=>'Ruby'
        }).
          to_return(status: 200, body: SUCCESS_ADD_DOMAIN, headers: {})
  end

  def test_it_domain_list
    success_domain_list
    create_new_short_url(DEFAULT_HOST_NAME, DEFAULT_API_KEY)
    @short_url.domain_list
    assert_equal @short_url.domain_list, SUCCESS_DOMAIN_LIST
  end

  def success_domain_list
    stub_request(:get, "https://api.short.io/api/domains").
      with(
        headers: {
              'Accept'=>'application/json',
              'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
              'Authorization'=>'API_KEY',
              'Host'=>'api.short.io',
              'User-Agent'=>'Ruby'
        }).
          to_return(status: 200, body: SUCCESS_DOMAIN_LIST, headers: {})
  end

  def test_it_delete_domain
    success_delete_domain
    create_new_short_url(DEFAULT_HOST_NAME, DEFAULT_API_KEY, nil, { domain_id: 7252 })
    @short_url.delete_domain
    assert_equal @short_url.delete_domain, SUCCESS_DELETE_OR_UPDATE
  end

  def test_it_update_domain
    success_update_domain
    create_new_short_url(DEFAULT_HOST_NAME, DEFAULT_API_KEY, nil,{domain_id: 7252, root_redirect_url: 'https://short.cm'})
    @short_url.update_domain
    assert_equal @short_url.update_domain, SUCCESS_DELETE_OR_UPDATE
  end

  def success_delete_domain
    stub_request(:post, "https://api.short.io/domains/delete/").
      with(
        headers: {
              'Accept'=>'*/*',
              'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
              'Authorization'=>'API_KEY',
              'Host'=>'api.short.io',
              'User-Agent'=>'Ruby'
        }).
          to_return(status: 200, body: SUCCESS_DELETE_OR_UPDATE, headers: {})
  end

  def success_update_domain
    stub_request(:post, "https://api.short.io/domains/settings/").
      with(
        body: "{\"rootRedirect\":\"\"}",
        headers: {
              'Accept'=>'application/json',
              'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
              'Authorization'=>'API_KEY',
              'Content-Type'=>'application/json',
              'Host'=>'api.short.io',
              'User-Agent'=>'Ruby'
        }).
          to_return(status: 200, body: SUCCESS_DELETE_OR_UPDATE, headers: {})
  end

  def create_new_short_url(host_name=DEFAULT_HOST_NAME, api_key=DEFAULT_API_KEY, short_io_base_url=SHORT_IO_BASE_URL, options={})
    @short_url = ::ShortIo::ShortUrl.new(host_name, api_key, options)
  end

  def default_initialize(host_name=DEFAULT_HOST_NAME, api_key=DEFAULT_API_KEY, short_io_base_url=SHORT_IO_BASE_URL, options={})
    create_new_short_url(host_name, api_key, options)
    @short_url.yield_self.inspect.include? '@api_key'
    @short_url.yield_self.inspect.include? "#{api_key}"
    @short_url.yield_self.inspect.include? '@hide_referer'
    @short_url.yield_self.inspect.include? "#{options.key?(:hide_referer) ? options[:hide_referer] : false}"
    @short_url.yield_self.inspect.include? '@host_name'
    @short_url.yield_self.inspect.include? "#{host_name}"
    @short_url.yield_self.inspect.include? '@https_link'
    @short_url.yield_self.inspect.include? "#{options.key?(:https_link) ? options[:https_link] : false}"
    @short_url.yield_self.inspect.include? '@link_type'
    @short_url.yield_self.inspect.include? options.key?(:link_type) ? options[:link_type] : 'random'
    @short_url.yield_self.inspect.include? '@short_io_base_url'
    @short_url.yield_self.inspect.include? SHORT_IO_BASE_URL
  end
end


lib/short_io/short_url.rb
require 'uri'
require 'net/http'
require 'openssl'
require 'json'

# @author Yosef Benny Widyokarsono
module ShortIo
  # Add domain and list domains registerd to Short.io
  class ShortUrl
    REQUEST_TYPE          = 'application/json'
    SHORT_IO_BASE_URL     = 'https://api.short.io/domains/'
    SHORT_IO_API_BASE_URL = 'https://api.short.io/api/domains'

    attr_reader :host_name, :api_key, :options
    
    def initialize(host_name, api_key, short_io_base_url=SHORT_IO_BASE_URL, options={})
      @host_name          ||= host_name
      @api_key            ||= api_key
      @short_io_base_url  ||= short_io_base_url
      @options            ||= options
      unless options.key?(:domain_id)
        options_default_value
        check_variables
      end
    end

    # Options Default Value
    #
    # @return [Hash] return default options in Hash if no value provided.

    def options_default_value
      @options = {
        hide_referer: options.key?(:hide_referer) ? options[:hide_referer] : false,
        https_link: options.key?(:https_link) ? options[:https_link] : false,
        link_type: options.key?(:link_type) ? options[:link_type] : 'random'
      }
    end

    # Check Variables
    #
    # @raise [HostNameError] if no Host Name provided.
    # @raise [ApiKeyError] if no API Key provided.

    def check_variables
      raise ShortIo::HostNameError.new(host_name: 'Please provide a host name') if (@host_name.nil? || @host_name.empty?)
      raise ShortIo::ApiKeyError.new(api_key: 'Please provide an API key') if (@api_key.nil? || @api_key.empty?)
    end

    # Prepare request

    def setup(action=nil, base_url=SHORT_IO_BASE_URL)
      @url = action.nil? ? URI(base_url) : URI(base_url + action)
      @http = Net::HTTP.new(@url.host, @url.port)
      @http.use_ssl = true
      @http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end

    # Add new domain
    #
    # @return [JSON] return in JSON format.
    #
    # @example
    #   ShortIo::ShortUrl.new('example.com', 'YOUR_API_KEY').add_domain
    #     {
    #       linkType: 'random',
    #       state: 'configured',
    #       cloaking: false,
    #       setupType: 'dns',
    #       httpsLinks: false,
    #       id: 91576,
    #       hostname: 'yourdomain.com',
    #       UserId: 9346,
    #       updatedAt: '2022-02-03T10:22:47.010Z',
    #       createdAt: '2022-02-03T10:22:46.649Z',
    #       provider: null,
    #       unicodeHostname: 'urdomain.com',
    #       clientStorage: null
    #     }
    #
    # @see https://developers.short.io/docs/adding-a-domain

    def add_domain
      setup

      request = Net::HTTP::Post.new(@url)
      request["accept"] = REQUEST_TYPE
      request["content-type"] = REQUEST_TYPE
      request["authorization"] = @api_key
      request.body = JSON.generate(
        {
          "hideReferer":"#{@options[:hide_referer]}",
          "httpsLinks":"#{@options[:https_link]}",
          "hostname":"#{@host_name}",
          "linkType":"#{@options[:link_type]}"
        }
      )
      response = @http.request(request)
      return response.read_body
    end

    # Delete a domain
    #
    # @return [JSON] return in JSON format.
    #
    # @example
    #   ShortIo::ShortUrl.new('example.com', 'YOUR_API_KEY', {domain_id: YOUR_DOMAIN_ID}).delete_domain
    #     response:
    #     {
    #       success: true
    #     }
    #
    # @see https://developers.short.io/docs/deleting-a-domain

    def delete_domain
      setup("delete/#{@options[:domain_id]}")

      request = Net::HTTP::Post.new(@url)
      request["authorization"] = @api_key

      response = @http.request(request)
      return response.read_body
    end

    # Udpadating a domain
    #
    # @return [JSON] return in JSON format.
    #
    # @example
    #   ShortIo::ShortUrl.new('example.com', 'YOUR_API_KEY', {domain_id: YOUR_DOMAIN_ID, root_redirect_url: 'YOUR_LINK'}).update_domain
    #     response:
    #     {
    #       success: true
    #     }
    #
    # @see https://developers.short.io/docs/updating-a-domain

    def update_domain
      setup("settings/#{@options[:domain_id]}")

      request = Net::HTTP::Post.new(@url)
      request["accept"] = REQUEST_TYPE
      request["content-type"] = REQUEST_TYPE
      request["authorization"] = @api_key
      request.body = JSON.generate({ 'rootRedirect': "#{@options[:root_redirect_url]}" })

      response = @http.request(request)
      return response.read_body
    end

    # Domain List
    #
    # @return [JSON] return domain list in JSON format.
    #
    # @example
    #   ShortIo::ShortUrl.new('example.com', 'YOUR_API_KEY').domain_list
    #     response:
    #     {
    #       id: 7252,
    #      hostname: 'yrbrand.co',
    #      title: null,
    #      segmentKey: null,
    #      linkType: 'increment',
    #      state: 'not_configured',
    #      provider: 'cloudflare',
    #      redirect404: 'https://short.cm',
    #      hideReferer: 1,
    #      caseSensitive: true,
    #      exportEnabled: true,
    #      cloaking: false,
    #      jsRedir: true,
    #      incrementCounter: 'A',
    #      setupType: 'js',
    #      autodeletePeriod: 1,
    #      httpsLinks: true,
    #      clientStorage: '{"configurationHidden":false}',
    #      integrationGA: null,
    #      integrationFB: null,
    #      integrationAdroll: null,
    #      integrationGTM: null,
    #      createdAt: '2017-12-07T08:24:41.000Z',
    #      updatedAt: '2019-12-24T13:08:30.000Z',
    #      TeamId: 1381,
    #      unicodeHostname: 'yrbrand.co'
    #    }
    #
    # @see https://developers.short.io/docs/getting-a-list-of-domains
    
    def domain_list
      setup(nil, SHORT_IO_API_BASE_URL)
      request = Net::HTTP::Get.new(@url)
      request["accept"] = REQUEST_TYPE
      request["authorization"] = @api_key
      
      response = @http.request(request)
      return response.read_body
    end
  end
end